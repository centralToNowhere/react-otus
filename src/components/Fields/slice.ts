import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { fieldHeight, fieldWidth } from "@/utils";
import { getDataFromStorage } from "@/storage";

export interface IFieldControlState {
  cellSize: number;
  maxFieldWidth: number;
  maxFieldHeight: number;
  capacity: number;
  speed: number;
}

export const defaultFieldControlState: IFieldControlState = {
  cellSize: 40,
  maxFieldWidth: fieldWidth,
  maxFieldHeight: fieldHeight,
  capacity: 50,
  speed: 2,
};

const storageData = getDataFromStorage()?.fieldControl;

export const initialState: IFieldControlState = {
  cellSize: storageData?.cellSize || defaultFieldControlState.cellSize,
  maxFieldWidth:
    storageData?.maxFieldWidth || defaultFieldControlState.maxFieldWidth,
  maxFieldHeight:
    storageData?.maxFieldHeight || defaultFieldControlState.maxFieldHeight,
  capacity: storageData?.capacity || defaultFieldControlState.capacity,
  speed: storageData?.speed || defaultFieldControlState.speed,
};

export const fieldControlSlice = createSlice({
  name: "fieldControl",
  initialState,
  reducers: {
    setCellSize: (state, action: PayloadAction<number>) => {
      state.cellSize = action.payload;
    },
    setMaxFieldWidth: (state, action: PayloadAction<number>) => {
      state.maxFieldWidth = action.payload;
    },
    setMaxFieldHeight: (state, action: PayloadAction<number>) => {
      state.maxFieldHeight = action.payload;
    },
    setCapacity: (state, action: PayloadAction<number>) => {
      state.capacity = action.payload;
    },
    setSpeed: (state, action: PayloadAction<number>) => {
      state.speed = action.payload;
    },
    resetFieldControls: (state) => {
      state = defaultFieldControlState;
      return state;
    },
  },
});

export const {
  setCellSize,
  setMaxFieldWidth,
  setMaxFieldHeight,
  setCapacity,
  setSpeed,
  resetFieldControls,
} = fieldControlSlice.actions;
