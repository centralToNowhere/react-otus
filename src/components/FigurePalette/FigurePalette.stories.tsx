import React from "react";
import { Story, Meta } from "@storybook/react";
import {
  FigurePalette,
  FigurePaletteProps,
} from "@/components/FigurePalette/FigurePalette";

export default {
  title: "Components/FigurePalette",
  component: FigurePalette,
  args: {
    figures: [
      {
        name: "First figure",
        indexedCells: [
          1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0,
          0, 1,
        ],
        cellsInRow: 5,
        cellsInCol: 5,
      },
      {
        name: "Second figure",
        indexedCells: [1, 0, 1, 0, 1, 1, 0, 0, 0],
        cellsInRow: 3,
        cellsInCol: 3,
      },
      {
        name: "Third figure",
        indexedCells: [
          1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1,
          0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1,
          0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1,
          0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0,
          0, 0, 1, 1, 0, 0, 0, 1,
        ],
        cellsInRow: 10,
        cellsInCol: 10,
      },
    ],
  },
} as Meta;

const Template: Story<FigurePaletteProps> = (args) => (
  <FigurePalette {...args} />
);

export const PaletteDefault = Template;
