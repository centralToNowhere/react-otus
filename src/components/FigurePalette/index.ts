export {
  FigurePalette,
  paletteActiveCancelText,
  paletteActiveText,
} from "./FigurePalette";
export { FigurePaletteContainer } from "./FigurePaletteContainer";
export {
  figurePaletteSlice,
  setFigures,
  setFigurePaletteActive,
  setCurrentFigureIndex,
  resetPalette,
} from "./slice";
export { selectFigures } from "./selectors";
export type { IFigurePaletteState } from "./slice";
export type { CellFigure } from "@/components/FigurePalette/FigurePaletteContainer";
