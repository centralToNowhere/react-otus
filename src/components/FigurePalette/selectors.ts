import { RootState } from "@/store/redux/store";
import { createSelector } from "@reduxjs/toolkit";

export const selectFigurePalette = (state: RootState) => {
  return state.figurePalette;
};

export const selectFigures = createSelector(
  selectFigurePalette,
  (figurePalette) => {
    return figurePalette.figures;
  }
);

export const selectPaletteActive = createSelector(
  selectFigurePalette,
  (figurePalette) => {
    return figurePalette.figurePaletteActive;
  }
);

export const selectPaletteCurrent = createSelector(
  selectFigurePalette,
  (figurePalette) => {
    return figurePalette.currentFigureIndex;
  }
);
