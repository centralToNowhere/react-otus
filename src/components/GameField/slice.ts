import { ICell } from "@/Cell/Cell";
import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { getInitialCells } from "@/utils/CellGenerator";
import { initialState as FieldControlInitialState } from "@/components/Fields/slice";
import { getDataFromStorage } from "@/storage";

export interface IGameFieldState {
  activeCells: ICell[];
  gameInProgress: boolean;
}

export const defaultGameFieldState: IGameFieldState = {
  activeCells: [],
  gameInProgress: false,
};

const storageData = getDataFromStorage();

const w =
  storageData?.fieldControl?.maxFieldWidth ||
  FieldControlInitialState.maxFieldWidth;
const h =
  storageData?.fieldControl?.maxFieldHeight ||
  FieldControlInitialState.maxFieldHeight;
const cellSize =
  storageData?.fieldControl?.cellSize || FieldControlInitialState.cellSize;

const initialCells = getInitialCells(w, h, cellSize);

const initialState: IGameFieldState = {
  activeCells: initialCells,
  gameInProgress: false,
};

export const gameFieldSlice = createSlice({
  name: "gameField",
  initialState,
  reducers: {
    setActiveCells: (state, action: PayloadAction<ICell[]>) => {
      state.activeCells = action.payload;
    },
    setActiveCell: (state, action: PayloadAction<ICell>) => {
      state.activeCells = [...state.activeCells, action.payload];
    },
    setInactiveCell: (state, action: PayloadAction<ICell>) => {
      state.activeCells = state.activeCells.filter((cell: ICell) => {
        return cell.x !== action.payload.x || cell.y !== action.payload.y;
      });
    },
    resetCells: (state) => {
      state = defaultGameFieldState;
      return state;
    },
    startGame: (state) => {
      state.gameInProgress = true;
    },
    stopGame: (state) => {
      state.gameInProgress = false;
    },
    generateRandom: (state) => {
      return state;
    },
  },
});

export const {
  setActiveCells,
  setActiveCell,
  setInactiveCell,
  resetCells,
  startGame,
  stopGame,
  generateRandom,
} = gameFieldSlice.actions;
