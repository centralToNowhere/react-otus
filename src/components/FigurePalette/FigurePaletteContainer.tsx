import React from "react";
import { IndexedCells } from "@/Cell/Cell";
import { useDispatch, useSelector } from "react-redux";
import {
  selectFigures,
  selectPaletteActive,
} from "@/components/FigurePalette/selectors";
import { FigurePalette } from "@/components/FigurePalette/FigurePalette";
import {
  setCurrentFigureIndex,
  setFigurePaletteActive,
} from "@/components/FigurePalette/slice";

export interface CellFigure {
  name: string;
  indexedCells: IndexedCells;
  cellsInCol: number;
  cellsInRow: number;
  cellSize?: number;
}

export const FigurePaletteContainer = () => {
  const dispatch = useDispatch();
  const figures = useSelector(selectFigures);
  const paletteActive = useSelector(selectPaletteActive);

  const onStartFigurePlacement = () => {
    dispatch(setFigurePaletteActive(true));
  };

  const onCancelFigurePlacement = () => {
    dispatch(setFigurePaletteActive(false));
  };

  const setCurrentFigure = (index: number) => {
    dispatch(setCurrentFigureIndex(index));
  };

  return (
    <FigurePalette
      figures={figures}
      paletteActive={paletteActive}
      onStartFigurePlacement={onStartFigurePlacement}
      onCancelFigurePlacement={onCancelFigurePlacement}
      setCurrentFigureIndex={setCurrentFigure}
    />
  );
};
