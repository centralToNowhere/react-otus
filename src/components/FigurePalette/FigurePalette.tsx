import React, { FC, SVGAttributes, useState } from "react";
import styled from "@emotion/styled";
import { COLORS } from "@/styles/ui-styled";
import { Main as GameField } from "@/components/GameField/GameField";
import { FormButton } from "@/components/Buttons";
import { CellFigure } from "@/components/FigurePalette";

export interface FigurePaletteProps {
  figures: CellFigure[];
}

export const FigurePalette: FC<FigurePaletteProps> = ({ figures }) => {
  const [currentFigureIndex, setCurrentFigureIndex] = useState<number>(0);

  return (
    <FigurePaletteContainer data-testid={"figurePalette"}>
      <FigureName>{figures[currentFigureIndex].name}</FigureName>
      <FigureSliderContainer>
        <FigureSliderArrow
          onClick={() => {
            setCurrentFigureIndex((prev) => {
              return prev > 0 ? prev - 1 : prev;
            });
          }}
          viewBox="0 0 40 60"
          xmlns="http://www.w3.org/2000/svg"
          disabled={currentFigureIndex === 0}
          role={"button"}
          aria-label={"prev"}
        >
          <path d="M40 0 L40 60 L0 30 Z" />
        </FigureSliderArrow>
        <PaletteGameField>
          <GameField
            cellSize={20}
            indexedCells={figures[currentFigureIndex].indexedCells}
            cellsInRow={figures[currentFigureIndex].cellsInRow}
            cellsInCol={figures[currentFigureIndex].cellsInCol}
            setActiveCell={() => {
              // empty
            }}
            setInactiveCell={() => {
              // empty
            }}
          />
        </PaletteGameField>
        <FigureSliderArrow
          onClick={() => {
            setCurrentFigureIndex((prev) => {
              return prev !== figures.length - 1 ? prev + 1 : prev;
            });
          }}
          viewBox="0 0 40 60"
          xmlns="http://www.w3.org/2000/svg"
          disabled={currentFigureIndex === figures.length - 1}
          role={"button"}
          aria-label={"next"}
        >
          <path d="M0 0 L40 30 L0 60 Z" />
        </FigureSliderArrow>
      </FigureSliderContainer>
      <FormButton>Выбрать</FormButton>
    </FigurePaletteContainer>
  );
};

export const FigureName = styled.p`
  margin-bottom: 1rem;
  text-align: center;
`;

export const PaletteGameField = styled.div`
  margin: 1rem 2rem;
`;

export const FigurePaletteContainer = styled.div`
  display: inline-flex;
  flex-direction: column;
  max-width: 600px;
`;

export const FigureSliderArrow = styled.svg<
  SVGAttributes<SVGElement> & {
    disabled: boolean;
  }
>`
  width: 1rem;
  cursor: pointer;

  & path {
    fill: ${({ disabled }) => (disabled ? COLORS.border : COLORS.accent)};
  }
`;

export const FigureSliderContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
`;

export const FigureSlider = styled.div`
  display: flex;
`;
