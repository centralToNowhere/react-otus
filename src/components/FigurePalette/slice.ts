import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { CellFigure } from "@/components/FigurePalette/FigurePaletteContainer";

export interface IFigurePaletteState {
  figures: CellFigure[];
  figurePaletteActive: boolean;
  currentFigureIndex: number;
}

export const initialState: IFigurePaletteState = {
  figures: [
    {
      name: "Glider",
      indexedCells: [0, 1, 0, 0, 0, 1, 1, 1, 1],
      cellsInRow: 3,
      cellsInCol: 3,
    },
    {
      name: "Penta-decathlon",
      indexedCells: [
        0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0,
        0, 0, 0, 1, 0, 0,
      ],
      cellsInRow: 10,
      cellsInCol: 3,
    },
    {
      name: "Gosper glider gun",
      indexedCells: [
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
        0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1,
        1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
        0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
        1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
        0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
      ],
      cellsInRow: 36,
      cellsInCol: 9,
      cellSize: 10,
    },
  ],
  figurePaletteActive: false,
  currentFigureIndex: 0,
};

export const figurePaletteSlice = createSlice({
  name: "figurePalette",
  initialState,
  reducers: {
    setFigures: (state, action: PayloadAction<CellFigure[]>) => {
      state.figures = action.payload;
    },
    setFigurePaletteActive: (state, action: PayloadAction<boolean>) => {
      state.figurePaletteActive = action.payload;
    },
    setCurrentFigureIndex: (state, action: PayloadAction<number>) => {
      state.currentFigureIndex = action.payload;
    },
  },
});

export const { setFigures, setFigurePaletteActive, setCurrentFigureIndex } =
  figurePaletteSlice.actions;
