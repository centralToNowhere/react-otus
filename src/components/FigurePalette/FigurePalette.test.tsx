import React from "react";
import {
  FigurePaletteContainer,
  paletteActiveCancelText,
  paletteActiveText,
} from "@/components/FigurePalette";
import { render, screen, within, waitFor } from "@/utils/test-utils";
import { IFigurePaletteState } from "@/components/FigurePalette/slice";
import { COLORS } from "@/styles/ui-styled";
import userEvent from "@testing-library/user-event/dist";

const initialState: {
  figurePalette: IFigurePaletteState;
} = {
  figurePalette: {
    figures: [
      {
        name: "First figure",
        indexedCells: [
          1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0,
          0, 1,
        ],
        cellsInRow: 5,
        cellsInCol: 5,
      },
      {
        name: "Second figure",
        indexedCells: [1, 0, 1, 0, 1, 1, 0, 0, 0],
        cellsInRow: 3,
        cellsInCol: 3,
      },
      {
        name: "Third figure",
        indexedCells: [
          1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1,
          0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1,
          0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1,
          0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0,
          0, 0, 1, 1, 0, 0, 0, 1,
        ],
        cellsInRow: 10,
        cellsInCol: 10,
      },
    ],
    figurePaletteActive: false,
    currentFigureIndex: 0,
  },
};

describe("FigurePalette tests", () => {
  it("should render FigurePalette", () => {
    const { asFragment } = render(<FigurePaletteContainer />, {
      preloadedState: initialState,
    });

    const PaletteContainer = screen.getByTestId("figurePalette");
    const FigureName = screen.getByText("First figure");
    const PaletteStateButton = screen.getByText(paletteActiveText);
    const RotationButton = screen.getByLabelText("rotate");
    const PaletteGameField = within(PaletteContainer).getByTestId("field");
    const PrevArrow = within(PaletteContainer).getByLabelText("prev");
    const NextArrow = within(PaletteContainer).getByLabelText("next");

    expect(asFragment()).toMatchSnapshot();
    expect(PaletteContainer).toBeInTheDocument();
    expect(FigureName).toBeInTheDocument();
    expect(PaletteStateButton).toBeInTheDocument();
    expect(RotationButton).toBeInTheDocument();
    expect(PaletteGameField).toBeInTheDocument();
    expect(PrevArrow).toBeInTheDocument();
    expect(NextArrow).toBeInTheDocument();
  });

  it("should render 5 x 5 figure inside palette", () => {
    render(<FigurePaletteContainer />, {
      preloadedState: initialState,
    });

    const PaletteContainer = screen.getByTestId("figurePalette");
    const PaletteGameField = within(PaletteContainer).getByTestId("field");
    const cells: HTMLElement[] =
      within(PaletteGameField).getAllByTestId("cell");

    expect(cells.length).toBe(25);

    const indexedCells = initialState.figurePalette.figures[0].indexedCells;

    cells.forEach((cell, i) => {
      if (indexedCells[i]) {
        expect(cell).toHaveStyle({
          background: COLORS.activeCellBg,
        });
      } else {
        expect(cell).not.toHaveStyle({
          background: COLORS.activeCellBg,
        });
      }
    });
  });

  it("should not switch prev figure", async () => {
    render(<FigurePaletteContainer />, {
      preloadedState: initialState,
    });

    const PaletteContainer = screen.getByTestId("figurePalette");
    const PrevArrow = within(PaletteContainer).getByLabelText("prev");
    const PaletteGameField = within(PaletteContainer).getByTestId("field");

    expect(within(PaletteGameField).getAllByTestId("cell").length).toBe(25);
    userEvent.click(PrevArrow);

    await new Promise((resolve) => {
      setTimeout(() => {
        resolve(true);
      }, 100);
    });

    expect(within(PaletteGameField).getAllByTestId("cell").length).toBe(25);
  });

  it("should not switch to next figure", async () => {
    render(<FigurePaletteContainer />, {
      preloadedState: {
        ...initialState,
        figurePalette: {
          figures: [
            {
              name: "First figure",
              indexedCells: [
                1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1,
                0, 0, 0, 1,
              ],
              cellsInRow: 5,
              cellsInCol: 5,
            },
          ],
          currentFigureIndex: 0,
          figurePaletteActive: false,
        },
      },
    });

    const PaletteContainer = screen.getByTestId("figurePalette");
    const NextArrow = within(PaletteContainer).getByLabelText("next");
    const PaletteGameField = within(PaletteContainer).getByTestId("field");

    expect(within(PaletteGameField).getAllByTestId("cell").length).toBe(25);
    userEvent.click(NextArrow);

    await new Promise((resolve) => {
      setTimeout(() => {
        resolve(true);
      }, 100);
    });

    expect(within(PaletteGameField).getAllByTestId("cell").length).toBe(25);
  });

  it("should switch to next figure", async () => {
    render(<FigurePaletteContainer />, {
      preloadedState: initialState,
    });

    const PaletteContainer = screen.getByTestId("figurePalette");
    const NextArrow = within(PaletteContainer).getByLabelText("next");
    const PaletteGameField = within(PaletteContainer).getByTestId("field");

    expect(within(PaletteGameField).getAllByTestId("cell").length).toBe(25);
    userEvent.click(NextArrow);

    await new Promise((resolve) => {
      setTimeout(() => {
        resolve(true);
      }, 100);
    });

    expect(within(PaletteGameField).getAllByTestId("cell").length).toBe(9);
    userEvent.click(NextArrow);

    await new Promise((resolve) => {
      setTimeout(() => {
        resolve(true);
      }, 100);
    });

    expect(within(PaletteGameField).getAllByTestId("cell").length).toBe(100);
  });

  it("should switch to prev figure", async () => {
    render(<FigurePaletteContainer />, {
      preloadedState: initialState,
    });

    const PaletteContainer = screen.getByTestId("figurePalette");
    const NextArrow = within(PaletteContainer).getByLabelText("next");
    const PrevArrow = within(PaletteContainer).getByLabelText("prev");
    const PaletteGameField = within(PaletteContainer).getByTestId("field");

    userEvent.click(NextArrow);
    userEvent.click(NextArrow);

    await new Promise((resolve) => {
      setTimeout(() => {
        resolve(true);
      }, 100);
    });

    expect(within(PaletteGameField).getAllByTestId("cell").length).toBe(100);
    userEvent.click(PrevArrow);

    await new Promise((resolve) => {
      setTimeout(() => {
        resolve(true);
      }, 100);
    });

    expect(within(PaletteGameField).getAllByTestId("cell").length).toBe(9);
    userEvent.click(PrevArrow);

    await new Promise((resolve) => {
      setTimeout(() => {
        resolve(true);
      }, 100);
    });

    expect(within(PaletteGameField).getAllByTestId("cell").length).toBe(25);
  });

  it("should change palette button text to cancel state", () => {
    render(<FigurePaletteContainer />, {
      preloadedState: initialState,
    });

    const paletteStateButton = screen.getByText(paletteActiveText);

    userEvent.click(paletteStateButton);

    expect(paletteStateButton).toHaveTextContent(paletteActiveCancelText);
  });

  it("should change palette button text to active", () => {
    render(<FigurePaletteContainer />, {
      preloadedState: {
        ...initialState,
        figurePalette: {
          ...initialState.figurePalette,
          figurePaletteActive: true,
        },
      },
    });

    const paletteStateButton = screen.getByText(paletteActiveCancelText);

    userEvent.click(paletteStateButton);

    expect(paletteStateButton).toHaveTextContent(paletteActiveText);
  });

  it("should rotate figure on rotate button click", async () => {
    const { store } = render(<FigurePaletteContainer />, {
      preloadedState: {
        ...initialState,
        figurePalette: {
          ...initialState.figurePalette,
          figures: [
            {
              name: "Figure",
              indexedCells: [1, 0, 1, 0, 1, 1, 0, 0, 0],
              cellsInRow: 3,
              cellsInCol: 3,
            },
          ],
          currentFigureIndex: 0,
        },
      },
    });

    const rotations = [
      [1, 1, 0, 0, 1, 0, 1, 0, 0],
      [0, 0, 0, 1, 1, 0, 1, 0, 1],
      [0, 0, 1, 0, 1, 0, 0, 1, 1],
      [1, 0, 1, 0, 1, 1, 0, 0, 0],
    ];

    const rotateButton = screen.getByLabelText("rotate");

    await rotations.reduce((promise, rotation) => {
      return promise.then(() => {
        userEvent.click(rotateButton);

        return waitFor(() => {
          expect(store.getState().figurePalette).toEqual(
            expect.objectContaining({
              figures: [
                {
                  name: "Figure",
                  indexedCells: rotation,
                  cellsInRow: 3,
                  cellsInCol: 3,
                },
              ],
            })
          );
        });
      });
    }, Promise.resolve());
  });
});
