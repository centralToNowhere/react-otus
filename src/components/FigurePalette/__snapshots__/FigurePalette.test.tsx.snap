// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FigurePalette tests should render FigurePalette 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  max-width: 600px;
}

.emotion-2 {
  margin-bottom: 1rem;
  text-align: center;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-6 {
  width: 1rem;
  cursor: pointer;
}

.emotion-6 path {
  fill: #6a6d75;
}

.emotion-8 {
  margin: 1rem 2rem;
}

.emotion-10 {
  width: 100px;
  cursor: pointer;
  height: 100px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  background: rgb(193,184,244);
  margin: 0 auto;
}

.emotion-10 .cell {
  -webkit-transition: 0.5s ease background;
  transition: 0.5s ease background;
  width: 20px;
  height: 20px;
  line-height: 20px;
  text-align: center;
  border: 1px solid #e8d5f0;
}

.emotion-10 .cell-active {
  background: rgb(144, 131, 230);
}

.emotion-12 {
  width: 1rem;
  cursor: pointer;
}

.emotion-12 path {
  fill: rgb(255, 33, 92);
}

.emotion-14 {
  background: inherit;
  border-color: #6a6d75;
  line-height: 1.5;
}

.emotion-14:focus {
  outline: 2px solid rgb(255, 33, 92)!important;
  border: 2px solid transparent;
}

<div
    class="emotion-0 emotion-1"
    data-testid="figurePalette"
  >
    <p
      class="emotion-2 emotion-3"
    >
      First figure
    </p>
    <div
      class="emotion-4 emotion-5"
    >
      <svg
        aria-label="prev"
        class="emotion-6 emotion-7"
        disabled=""
        role="button"
        viewBox="0 0 40 60"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M40 0 L40 60 L0 30 Z"
        />
      </svg>
      <div
        class="emotion-8 emotion-9"
      >
        <div
          class="game-field emotion-10 emotion-11"
          data-testid="field"
        >
          <div
            aria-label="1"
            class="cell cell-active"
            data-state="true"
            data-testid="cell"
          />
          <div
            aria-label="2"
            class="cell"
            data-state="false"
            data-testid="cell"
          />
          <div
            aria-label="3"
            class="cell"
            data-state="false"
            data-testid="cell"
          />
          <div
            aria-label="4"
            class="cell"
            data-state="false"
            data-testid="cell"
          />
          <div
            aria-label="5"
            class="cell cell-active"
            data-state="true"
            data-testid="cell"
          />
          <div
            aria-label="6"
            class="cell"
            data-state="false"
            data-testid="cell"
          />
          <div
            aria-label="7"
            class="cell cell-active"
            data-state="true"
            data-testid="cell"
          />
          <div
            aria-label="8"
            class="cell"
            data-state="false"
            data-testid="cell"
          />
          <div
            aria-label="9"
            class="cell cell-active"
            data-state="true"
            data-testid="cell"
          />
          <div
            aria-label="10"
            class="cell"
            data-state="false"
            data-testid="cell"
          />
          <div
            aria-label="11"
            class="cell"
            data-state="false"
            data-testid="cell"
          />
          <div
            aria-label="12"
            class="cell"
            data-state="false"
            data-testid="cell"
          />
          <div
            aria-label="13"
            class="cell cell-active"
            data-state="true"
            data-testid="cell"
          />
          <div
            aria-label="14"
            class="cell"
            data-state="false"
            data-testid="cell"
          />
          <div
            aria-label="15"
            class="cell"
            data-state="false"
            data-testid="cell"
          />
          <div
            aria-label="16"
            class="cell"
            data-state="false"
            data-testid="cell"
          />
          <div
            aria-label="17"
            class="cell cell-active"
            data-state="true"
            data-testid="cell"
          />
          <div
            aria-label="18"
            class="cell"
            data-state="false"
            data-testid="cell"
          />
          <div
            aria-label="19"
            class="cell cell-active"
            data-state="true"
            data-testid="cell"
          />
          <div
            aria-label="20"
            class="cell"
            data-state="false"
            data-testid="cell"
          />
          <div
            aria-label="21"
            class="cell cell-active"
            data-state="true"
            data-testid="cell"
          />
          <div
            aria-label="22"
            class="cell"
            data-state="false"
            data-testid="cell"
          />
          <div
            aria-label="23"
            class="cell"
            data-state="false"
            data-testid="cell"
          />
          <div
            aria-label="24"
            class="cell"
            data-state="false"
            data-testid="cell"
          />
          <div
            aria-label="25"
            class="cell cell-active"
            data-state="true"
            data-testid="cell"
          />
        </div>
      </div>
      <svg
        aria-label="next"
        class="emotion-12 emotion-7"
        role="button"
        viewBox="0 0 40 60"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0 L40 30 L0 60 Z"
        />
      </svg>
    </div>
    <button
      class="emotion-14 emotion-15"
    >
      Выбрать
    </button>
  </div>
</DocumentFragment>
`;
